doc/kernel/usermode/syscalls.rst

```diff
@@ -0,0 +1,259 @@
+  will be invoked instead.
+
+The body of the API is created in the generated system header. Using the
+example of :c:func:`k_sem_init()`, this API is declared in
+``include/kernel.h``. At the bottom of ``include/kernel.h`` is:
+
+::
+
+    #include <syscalls/kernel.h>
+
+Inside this header is the body of :c:func:``k_sem_init()``:
+
+::
+
+    K_SYSCALL_DECLARE3_VOID(K_SYSCALL_K_SEM_INIT, k_sem_init, struct k_sem *,
+                            sem, unsigned int, initial_count,
+                            unsigned int, limit);
+
+This generates an inline function which takes three arguments with void
+return value. Depending on context it will either directly call the
+implementation function or go through a system call elevation. A
+prototype for the implementation function is also automatically generated.
+
+The header containing :c:macro:`K_SYSCALL_DECLARE3_VOID()` is itself
+generated due to its repetitive nature and can be found in
+``include/generated/syscall_macros.h``.
+
+Implementation Function
+=======================
+
+The implementation function is what actually does the work for the API.
+Zephyr normally does little to no error checking of arguments, or does this
+kind of checking with assertions. When writing the implementation function,
+assume that all parameters have been validated, or check with assertions.
+
+All implementation functions must follow a naming convention, it is the
+name of the API prefixed with ``_impl_``. Implementation functions may be
+declared in the same header as the API as a static inline function or
+declared in some C file. There is no prototype needed for implementation
+functions, these are automatically generated.
+
+Handler Function
+================
+
+The handler function is when runs on the kernel side when a user thread makes
+a system call. When the user thread makes a software interrupt to elevate to
+supervisor mode, the common system call entry point uses the system call ID
+provided by the user to look up the appropriate handler function for that
+system call and jump into it.
+
+Handler functions only run when system call APIs are invoked from user mode.
+If an API is invoked from supervisor mode, the implementation is simply called.
+
+The purpose of the handler function is to validate all the arguments passed in.
+This includes:
+
+* Any kernel object pointers provided. For example, the semaphore APIs must
+  ensure that the semaphore object passed in is a valid semaphore and that
+  the calling thread has permission on it.
+
+* Any memory buffers passed in from user mode. Checks must be made that the
+  calling thread has read or write permissions on the provided buffer.
+
+* Any other arguments that have a limited range of valid values.
+
+Handler functions involve a great deal of boilerplate code which has been
+made simpler by some macros in ``kernel/include/syscall_handlers.h``. It is
+strongly recommended that any handler function be declared using these macros.
```

Comment:
Better would be:

     Handler functions should be declared using these macros.